1. Import Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Charger les données
df = pd.read_csv("Building_Energy_Benchmarking_Data__2015-Present_20250302.csv")

# Afficher les premières lignes
df.head()

OSEBuildingID  DataYear                    BuildingName    BuildingType  \
0              1      2023            MAYFLOWER PARK HOTEL  NonResidential   
1              2      2023                 PARAMOUNT HOTEL  NonResidential   
2              3      2023  WESTIN HOTEL (Parent Building)  NonResidential   
3              5      2023                       HOTEL MAX  NonResidential   
4              8      2023           WARWICK SEATTLE HOTEL  NonResidential   

  TaxParcelIdentificationNumber         Address     City State  ZipCode  \
0                     659000030   405 OLIVE WAY  SEATTLE    WA    98101   
1                     659000220     724 PINE ST  SEATTLE    WA    98101   
2                     659000475    1900 5TH AVE  SEATTLE    WA    98101   
3                     659000640  620 STEWART ST  SEATTLE    WA    98101   
4                     659000970   401 LENORA ST  SEATTLE    WA    98121   

   Latitude  ...  Electricity(kWh) SteamUse(kBtu)  NaturalGas(therms)  \
0  47.61220  ...         1047120.0      2797857.0             15998.0   
1  47.61307  ...          786237.0            NaN             38570.0   
2  47.61367  ...        11121130.0     20494222.0             59006.0   
3  47.61412  ...          780180.0      1248616.0             11231.0   
4  47.61375  ...         1401668.0            NaN             80926.0   

   ComplianceStatus                    ComplianceIssue  Electricity(kBtu)  \
0         Compliant                           No Issue          3572773.0   
1         Compliant                           No Issue          2682640.0   
2         Compliant                           No Issue         37945296.0   
3         Compliant  Missing EUI, Incomplete Fuel Data          2661973.0   
4         Compliant                           No Issue          4782491.0   

   NaturalGas(kBtu)  TotalGHGEmissions  GHGEmissionsIntensity  Demolished  
0         1599760.0              332.1                   3.76       False  
1         3856970.0              220.0                   2.49       False  
2         5900600.0             2192.8                   2.89       False  
3         1123100.0              176.1                   2.87       False  
4         8092580.0              456.6                   4.02       False  

[5 rows x 46 columns]

2. Observe Data

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# Vérification des dimensions du dataset 
print(f"Shape du dataset (nombre de lignes, nombre de colonnes) : {df.shape}")

# Vérification des noms des colonnes 
print("\nNoms des colonnes du dataset :")
print(df.columns.tolist())  # Affiche la liste des colonnes pour vérifier leur format

#  Aperçu rapide des données 
print("\n🔍 Aperçu des premières lignes :")
print(df.head())  # Afficher les 5 premières lignes

print("\n🔍 Aperçu des dernières lignes :")
print(df.tail())  # Afficher les 5 dernières lignes

print("\n🔍 Échantillon aléatoire de 5 lignes :")
print(df.sample(5))  # Afficher un échantillon aléatoire de 5 lignes

#  Informations générales sur les variables 
print("\n📌 Informations générales sur le dataset :")
df.info()  # Affiche le type des colonnes, le nombre de valeurs non nulles et la mémoire utilisée

#  Vérification des types de données 
print("\n📌 Types des données :")
print(df.dtypes)  # Vérifier si certaines colonnes doivent être converties (ex: dates, nombres mal formatés)

#  Vérification des valeurs manquantes 
print("\n⚠️ Nombre de valeurs manquantes par colonne :")
missing_values = df.isnull().sum()  # Compter les valeurs nulles par colonne
missing_values = missing_values[missing_values > 0]  # Filtrer uniquement les colonnes qui ont des valeurs manquantes
print(missing_values)


Shape du dataset (nombre de lignes, nombre de colonnes) : (30616, 46)

Noms des colonnes du dataset :
['OSEBuildingID', 'DataYear', 'BuildingName', 'BuildingType', 'TaxParcelIdentificationNumber', 'Address', 'City', 'State', 'ZipCode', 'Latitude', 'Longitude', 'Neighborhood', 'CouncilDistrictCode', 'YearBuilt', 'NumberofFloors', 'NumberofBuildings', 'PropertyGFATotal', 'PropertyGFABuildings', 'PropertyGFAParking', 'SelfReportGFATotal', 'SelfReportGFABuildings', 'SelfReportParking', 'ENERGYSTARScore', 'SiteEUIWN(kBtu/sf)', 'SiteEUI(kBtu/sf)', 'SiteEnergyUse(kBtu)', 'SiteEnergyUseWN(kBtu)', 'SourceEUIWN(kBtu/sf)', 'SourceEUI(kBtu/sf)', 'EPAPropertyType', 'LargestPropertyUseType', 'LargestPropertyUseTypeGFA', 'SecondLargestPropertyUseType', 'SecondLargestPropertyUseTypeGFA', 'ThirdLargestPropertyUseType', 'ThirdLargestPropertyUseTypeGFA', 'Electricity(kWh)', 'SteamUse(kBtu)', 'NaturalGas(therms)', 'ComplianceStatus', 'ComplianceIssue', 'Electricity(kBtu)', 'NaturalGas(kBtu)', 'TotalGHGEmi

Le dataset contient 30,616 lignes et 46 colonnes.
Plusieurs colonnes ont des valeurs manquantes, notamment :
- LargestPropertyUseTypeGFA (19,092 valeurs manquantes)
- SecondLargestPropertyUseTypeGFA (22,182 valeurs manquantes)
- Electricity(kWh), SteamUse(kBtu), NaturalGas(therms) (plusieurs milliers de valeurs manquantes)
- TotalGHGEmissions, GHGEmissionsIntensity (≈ 450 valeurs manquantes)

- Pour les colonnes très importantes (TotalGHGEmissions, GHGEmissionsIntensity...), on les traite maintenant.
- Pour les colonnes catégorielles (LargestPropertyUseTypeGFA, SecondLargestPropertyUseType...), on attend.

2.2. Observe data (cleaning missing values)

In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("Building_Energy_Benchmarking_Data__2015-Present_20250302.csv")

#  1. Vérification des valeurs manquantes 
print("\n⚠️ Nombre de valeurs manquantes par colonne :")
missing_values = df.isnull().sum()
missing_values = missing_values[missing_values > 0]  # Filtrer seulement les colonnes avec des valeurs manquantes
print(missing_values)

# Calcul du pourcentage de valeurs manquantes
missing_percentage = (df.isnull().sum() / len(df)) * 100
missing_percentage = missing_percentage[missing_percentage > 0]

# Afficher le pourcentage des valeurs manquantes
print("\n📊 Pourcentage de valeurs manquantes par colonne :")
print(missing_percentage.sort_values(ascending=False))

# 2. Suppression des colonnes avec trop de valeurs manquantes (si > 70%)
threshold = 70  # Seuil de pourcentage pour supprimer une colonne
cols_to_drop = missing_percentage[missing_percentage > threshold].index  # Colonnes à supprimer

df = df.drop(columns=cols_to_drop)  # Suppression des colonnes
print(f"\n🗑️ Colonnes supprimées : {list(cols_to_drop)}")
print(f"🔄 Nouvelle shape du dataset après suppression : {df.shape}")

# 3. Imputation des valeurs numériques manquantes #
# Colonnes numériques à remplir (en utilisant la médiane pour éviter les valeurs extrêmes)
num_cols = ['TotalGHGEmissions', 'GHGEmissionsIntensity', 'Electricity(kWh)', 'NaturalGas(therms)']
for col in num_cols:
    if col in df.columns:  # Vérifier si la colonne existe encore après suppression
        df[col].fillna(df[col].median(), inplace=True)

print("\n✅ Imputation des valeurs numériques avec la médiane terminée.")

# 4. Imputation des valeurs catégorielles manquantes 
# Colonnes catégorielles à remplir avec "Unknown"
cat_cols = ['LargestPropertyUseType', 'SecondLargestPropertyUseType']
for col in cat_cols:
    if col in df.columns:  # Vérifier si la colonne existe encore après suppression
        df[col].fillna("Unknown", inplace=True)

print("\n✅ Imputation des valeurs catégoriques terminée.")

# 5.  Vérification finale des valeurs manquantes 
print("\n🔍 Vérification finale des valeurs manquantes après traitement :")
print(df.isnull().sum()[df.isnull().sum() > 0])  # Vérifier s'il reste des valeurs manquantes



⚠️ Nombre de valeurs manquantes par colonne :
NumberofFloors                        83
NumberofBuildings                    294
ENERGYSTARScore                     7917
SiteEUIWN(kBtu/sf)                  1166
SiteEUI(kBtu/sf)                     856
SiteEnergyUse(kBtu)                  851
SiteEnergyUseWN(kBtu)               1161
SourceEUIWN(kBtu/sf)                1166
SourceEUI(kBtu/sf)                   856
LargestPropertyUseType                 8
LargestPropertyUseTypeGFA              3
SecondLargestPropertyUseType       13102
SecondLargestPropertyUseTypeGFA    13102
ThirdLargestPropertyUseType        23927
ThirdLargestPropertyUseTypeGFA     24088
Electricity(kWh)                     541
SteamUse(kBtu)                     29559
NaturalGas(therms)                 11448
ComplianceIssue                      345
Electricity(kBtu)                    541
NaturalGas(kBtu)                   11448
TotalGHGEmissions                    457
GHGEmissionsIntensity                459
dtype: int

/var/folders/y8/ghc5lkjj70vf0ggbg680pgsw0000gn/T/ipykernel_64276/2435877274.py:35: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(df[col].median(), inplace=True)
/var/folders/y8/ghc5lkjj70vf0ggbg680pgsw0000gn/T/ipykernel_64276/2435877274.py:35: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values alwa

1. Colonnes supprimées :
ThirdLargestPropertyUseType, ThirdLargestPropertyUseTypeGFA, et SteamUse(kBtu) ont bien été supprimées.
Nouvelle shape du dataset : (30616, 43) → Tout est bon.
2. Valeurs manquantes restantes :
Des valeurs sont encore manquantes pour ENERGYSTARScore (25%), NaturalGas(kBtu) (37%), etc.
Pour les autres, c'est beaucoup moins critique.

3.UnivariateAnalysis